functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,143 @@
+import torch
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from functorch.experimental.ops import PyOperator, fallthrough_fn
+from torch.utils._pytree import tree_flatten
+from torch.fx.experimental.proxy_tensor import get_isolated_graphmodule, get_proxy_slot
+import torch.utils._pytree as pytree
+from torch.utils._mode_utils import no_dispatch
+from torch.utils._python_dispatch import TorchDispatchMode
+import random
+import string
+
+"""
+We're going to define a `cond` operation.
+In order to do this, we need implementations for each of the dispatch keys.
+"""
+from contextlib import contextmanager
+
+# TODO(voz): Move out somewhere else once other py dispatched ops need it
+@contextmanager
+def suspend_mode(mode):
+    assert(mode is not None), "Cannot suspend None mode"
+    assert(isinstance(mode, TorchDispatchMode)), f"Unexpected mode type {mode.__class__}"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, args, kwargs=None):
+    assert kwargs is None or not kwargs
+    pred, true_fn, false_fn, operands = args
+
+    def _unwrap_proxy(e):
+        proxy_or_e = get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy )
+        return proxy_or_e
+
+    assert isinstance(operands, list), "Cond operands must be a list of tensors"
+    assert all(isinstance(o, torch.Tensor) for o in operands), "Cond operands must be a list of tensors"
+
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
```

Comment:
for full soundness, you must also test that the false graph variant also is unset

Reply:
The name testing strategy here means that you need to have copied all other submodules into the target GraphModule before you trace, otherwise the name selection here could clobber a name that would have been attempted to be used later in the trace.

Reply:
> for full soundness, you must also test that the false graph variant also is unset

Sure.

> The name testing strategy here means that you need to have copied all other submodules into the target GraphModule before you trace, otherwise the name selection here could clobber a name that would have been attempted to be used later in the trace.

Yes - is this an unsafe assumption?

Reply:
> Yes - is this an unsafe assumption?

Well, for `make_fx` you're safe because there is no starting module (and therefore no preexisting submodules). Of course, a user could do anything, include violate this invariant, but at least for the passes we have today I'm not aware of anyone violating this invariant.